src/ContractInteractor.ts

```diff
@@ -1,919 +1,618 @@
-                });
-
-            return {
-                verifierAccepted: true,
-                reverted: false,
-                returnValue: res.returnValue
-            };
-        } catch (e) {
-            const message =
-                e instanceof Error
-                    ? e.message
-                    : JSON.stringify(e, replaceErrors);
-            return {
-                verifierAccepted: true,
-                reverted: true,
-                returnValue: `view call to 'deployCall' reverted in client: ${message}`
-            };
-        }
-    }
-
-    async getMaxViewableDeployGasLimit(
-        request: DeployTransactionRequest
-    ): Promise<BN> {
-        const gasPrice = toBN(request.relayRequest.relayData.gasPrice);
-        let gasLimit = toBN(0);
-
-        if (!gasPrice.eq(toBN(0))) {
-            const maxEstimatedGas = toBN(
-                await this.walletFactoryEstimateGasOfDeployCall(request)
-            );
-            const workerBalanceAsUnitsOfGas = toBN(
-                await this.getBalance(
-                    request.relayRequest.relayData.relayWorker
-                )
-            ).div(gasPrice);
-
-            if (workerBalanceAsUnitsOfGas.gte(maxEstimatedGas)) {
-                gasLimit = maxEstimatedGas;
-            }
-        }
-
-        return gasLimit;
-    }
-
-    async estimateRelayTransactionMaxPossibleGas(
-        relayRequest: RelayRequest,
-        signature: PrefixedHexString
-    ): Promise<number> {
-        const maxPossibleGas = await this.estimateGas({
-            from: relayRequest.relayData.relayWorker,
-            to: relayRequest.request.relayHub,
-            data: this.relayHubInstance.contract.methods
-                .relayCall(relayRequest, signature)
-                .encodeABI(),
-            gasPrice: relayRequest.relayData.gasPrice
+  private readonly _provider: JsonRpcProvider;
+
+  public get provider(): JsonRpcProvider {
+    return this._provider;
+  }
+
+  private readonly _signer: Wallet;
+
+  public get signer(): Wallet {
+    return this._signer;
+  }
+
+  private readonly _config: EnvelopingConfig;
+
+  public get config(): EnvelopingConfig {
+    return this._config;
+  }
+
+  private readonly _versionManager: VersionsManager;
+
+  public get versionManager(): VersionsManager {
+    return this._versionManager;
+  }
+
+  public static async getInstance(
+    provider: JsonRpcProvider,
+    signer: Wallet,
+    config: EnvelopingConfig
+  ): Promise<ContractInteractor> {
+    if (!ContractInteractor.instance) {
+      ContractInteractor.instance = new ContractInteractor(
+        provider,
+        signer,
+        config
+      );
+      const version = await ContractInteractor.instance._relayHub.versionHub();
+      ContractInteractor.instance._validateVersion(version);
+    }
+
+    return ContractInteractor.instance;
+  }
+
+  private constructor(
+    provider: JsonRpcProvider,
+    signer: Wallet,
```

Comment:
The signer is mandatory. Most(Maybe all) of the interactions with the contracts are just querying data, Im not sure if its necessary to have a signer or with the provider will be enough.

Reply:
The signer is there (Wallet is a type of Signer), which actually make me realise that the provider is unnecessary, as signer contains provider.

Reply:
ah yea, it's there because the provider may be undefined in signer and because it has to be a rpc provider in order to get signer (though there might be a better way?).

Reply:
fixed